File: priority_queue/level2/BOJ_국회의원선거.py
import heapq
import sys

input = sys.stdin.readline


def solv():
    answer = 0
    N = int(input())
    dasom = 0
    max_heap = []
    for i in range(N):
        if i == 0:
            dasom = int(input())
        else:
            heapq.heappush(max_heap, -int(input()))
    while max_heap and dasom <= -max_heap[0]:
        mx = -heapq.heappop(max_heap)
        dasom += 1
        mx -= 1
        heapq.heappush(max_heap, -mx)
        answer += 1
    print(answer)
solv()

File: priority_queue/level2/BOJ_국회의원_선거.java
package priority_queue.level2;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Collections;
import java.util.PriorityQueue;

public class BOJ_국회의원_선거 {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuffer sb = new StringBuffer();
		int N = Integer.parseInt(br.readLine());
		int answer = 0;
		PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
		int dasom = Integer.parseInt(br.readLine());
		for (int i = 0; i < N - 1; i++) {
			maxHeap.add(Integer.parseInt(br.readLine()));
		}
		while (!maxHeap.isEmpty() && dasom <= maxHeap.peek()) {
			maxHeap.add(maxHeap.poll() - 1);
			dasom += 1;
			answer += 1;
		}
		System.out.println(answer);
	}
}
